src/components/image/index.tsx

```diff
@@ -46,44 +47,70 @@ export function Image({
   maxSize,
   alt,
   inputProps,
+  errorComponent: ErrorComponent,
   ...imgOptions
 }: Props): ReactElement {
+  const [error, setError] = useState<null | Error>(null);
   const { fetch } = useContext(SessionContext);
 
   const thingContext = useContext(ThingContext);
   const { thing: contextThing } = thingContext;
 
   const thing = propThing || contextThing;
 
-  if (!thing) {
-    throw new Error("Thing not found as property or in context");
+  if (!thing && !error) {
+    const thingNotFound = new Error(
+      "Thing not found as property or in context"
+    );
+
+    if (!ErrorComponent && !onError) {
+      throw thingNotFound;
+    } else if (onError) {
+      onError(thingNotFound);
+    }
+
+    setError(thingNotFound);
```

Comment:
I wonder if there's not a more ergonomic way to handle these repeated branches.

For example, you could maybe define such a function inside the component, after the hooks:

```typescript
function handleError(error: Error) {
    if (!ErrorComponent && !onError) {
      throw error;
    } else if (onError) {
      onError(error);
    }

    setError(error);
}
```

or perhaps delegate it to another component that you render instead?

Reply:
That would help, but I think I may save this refactoring until the below suggestion (pulling data fetching into a hook.)